### PR TITLE
feat: use secret store for PSK key (#33)

### PR DESCRIPTION
**Pull request by iain-anderson**
Signed-off-by: Iain Anderson <iain@iotechsys.com>

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-coap-c/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?) - no unit tests for C device services
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?) - documentation updated in this PR
  <link to docs PR>

## Testing Instructions

Run the service in PSK mode with a simulated device (see "Testing/Simulation for CoAP Server" in the README

### CODE REVIEW

**Comment by sudhamani-hcl** [comment]
LGTM

**Review by sudhamani-hcl — DISMISSED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -64,8 +63,24 @@ Below are the recognized properties for the Driver section, followed by an examp
...
+| Secret name | Value                                                                             |
+|-------------|-----------------------------------------------------------------------------------|
+| PskKey      | Pre-shared key. Accepts only a single key, ignored in NoSec mode.                 |
+
+For example if using insecure mode (secrets in configuration file):
+
+```
+  [Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.device-coap]
+      path = "device-coap"
```
Comment (lenny-intel) [inline README.md:80]: For consistency the path should be the name of the secret not the service name. Suggest using PSK. 
In secure mode the service should be part of the base path that this `path` is appended to it like:
```
.../device-coap/psk
```
`.../device-coap` is the secret store in Vault for Device Coap.
`.../device-coap/psk` is a Secret within the Device Coap secret store in vault.

The path in the InsucureSecrets needs to mimic that used for Vault.

File: src/c/device-coap.c [hunk 2]
```diff
@@ -47,34 +47,46 @@ coap_security_mode_t find_security_mode(const char *mode_text) {
...
+    case SECURITY_MODE_UNKNOWN: {
       iot_log_error(lc, "Unknown security mode");
       driver->psk_key = NULL;
-      return false;
+      result = false;
+      break;
+    }
 
     case SECURITY_MODE_PSK: {
+      iot_data_t *secrets = devsdk_get_secrets (driver->service, "device-coap");
```
Comment (lenny-intel) [inline src/c/device-coap.c:67]:
```suggestion
      iot_data_t *secrets = devsdk_get_secrets (driver->service, "psk");
```

File: res/configuration.toml [hunk 3]
```diff
@@ -52,3 +51,9 @@
     UpdateLastConnected = false
     [Writable.Device.Discovery]
     Enabled = false
+  [Writable.InsecureSecrets]
```
Comment (lenny-intel) [inline res/configuration.toml:54]: Also need the `SecretStore` cinfig section which should be:

```
[SecretStore]
Type = "vault"
Protocol = "http"
Host = "localhost"
Port = 8200
Path = "device-coap/"
TokenFile = "/tmp/edgex/secrets/device-coap/secrets-token.json"
RootCaCertPath = ""
ServerName = ""
  [SecretStore.Authentication]
  AuthType = "X-Vault-Token"
```
Note this is where the `device-coap` is added to the path.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Comment by iain-anderson** [comment]
updated with changes recommended by @lenny-intel

File: res/configuration-native.toml [hunk 4]
```diff
@@ -53,3 +52,21 @@
     UpdateLastConnected = false
     [Writable.Device.Discovery]
     Enabled = false
+  [Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.device-coap]
+      path = "device-coap"
+      [Writable.InsecureSecrets.device-coap.Secrets]
+        # Key is up to 16 arbitrary bytes; must be base64 encoded here
+        PskKey = "ME42aURHZ3Uva0Y0eG9lZw=="
```
Comment (lenny-intel) [inline res/configuration-native.toml:60]: Missing name changes that were applied in the README

File: res/configuration.toml [hunk 5]
```diff
@@ -52,3 +51,21 @@
...
+    [Writable.InsecureSecrets.device-coap]
+      path = "device-coap"
+      [Writable.InsecureSecrets.device-coap.Secrets]
```
Comment (lenny-intel) [inline res/configuration.toml:57]: Missing name changes that were applied in the README

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Pull request merged**